chain/actor.go

```diff
@@ -0,0 +1,240 @@
+	node          lens.API
+	opener        lens.APIOpener
+	closer        lens.APICloser
+	extractRaw    bool
+	extractParsed bool
+	lastTipSet    *types.TipSet
+	lastStateTree *state.StateTree
+}
+
+func NewActorStateProcessor(opener lens.APIOpener, extractRaw bool, extractParsed bool) *ActorStateProcessor {
+	p := &ActorStateProcessor{
+		opener:        opener,
+		extractRaw:    extractRaw,
+		extractParsed: extractParsed,
+	}
+	return p
+}
+
+func (p *ActorStateProcessor) ProcessTipSet(ctx context.Context, ts *types.TipSet) (model.PersistableWithTx, *visormodel.ProcessingReport, error) {
+	if p.node == nil {
+		node, closer, err := p.opener.Open(ctx)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("unable to open lens: %w", err)
+		}
+		p.node = node
+		p.closer = closer
+	}
+
+	var data model.PersistableWithTx
+	var report *visormodel.ProcessingReport
+	var err error
+
+	stateTree, err := state.LoadStateTree(p.node.Store(), ts.ParentState())
+	if err != nil {
+		return nil, nil, xerrors.Errorf("failed to load state tree: %w", err)
+	}
+
+	if p.lastTipSet != nil && p.lastStateTree != nil {
+		if p.lastTipSet.Height() > ts.Height() {
+			// last tipset seen was the child
+			data, report, err = p.processStateChanges(ctx, p.lastTipSet, ts, p.lastStateTree, stateTree)
+		} else if p.lastTipSet.Height() < ts.Height() {
+			// last tipset seen was the parent
+			data, report, err = p.processStateChanges(ctx, ts, p.lastTipSet, stateTree, p.lastStateTree)
+		} else {
+			log.Errorw("out of order tipsets", "height", ts.Height(), "last_height", p.lastTipSet.Height())
+		}
+	}
+
+	p.lastTipSet = ts
+	p.lastStateTree = stateTree
+
+	if err != nil {
+		log.Errorw("error received while processing actors, closing lens", "error", err)
+		if cerr := p.Close(); cerr != nil {
+			log.Errorw("error received while closing lens", "error", cerr)
+		}
+	}
+	return data, report, err
+}
+
+func (p *ActorStateProcessor) processStateChanges(ctx context.Context, ts *types.TipSet, pts *types.TipSet, stateTree *state.StateTree, parentStateTree *state.StateTree) (model.PersistableWithTx, *visormodel.ProcessingReport, error) {
+	log.Debugw("processing state changes", "height", ts.Height(), "parent_height", pts.Height())
+
+	report := &visormodel.ProcessingReport{
+		Height:    int64(ts.Height()),
+		StateRoot: ts.ParentState().String(),
+		Status:    visormodel.ProcessingStatusOK,
+	}
+
+	if !types.CidArrsEqual(ts.Parents().Cids(), pts.Cids()) {
+		report.ErrorsDetected = xerrors.Errorf("child is not on the same chain")
+		return nil, report, nil
+	}
+
+	changes, err := state.Diff(parentStateTree, stateTree)
+	if err != nil {
+		report.ErrorsDetected = xerrors.Errorf("failed to diff state trees: %w", err)
+		return nil, report, nil
+	}
+
+	ll := log.With("height", int64(ts.Height()))
+
+	ll.Debugw("found actor state changes", "count", len(changes))
+
+	start := time.Now()
+
+	// Run each task concurrently
+	results := make(chan *ActorStateResult, len(changes))
```

Comment:
what closes / cleans up this channel?

Reply:
It goes out of scope when the function returns and is garbage collected as normal. There's never any need to close a channel unless it is to signal that readers should stop waiting on it.